services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusUpdateRequest.kt

```diff
@@ -301,6 +301,54 @@ data class InvoicingPostalAddressRequest(
     val city: String?,
 )
 
+@JvmName("hasChangesInKaivuilmoitusAlueList")
+fun List<KaivuilmoitusAlue>?.hasChanges(areas: List<KaivuilmoitusAlue>?): Boolean {
+    if (this == null) {
+        return areas != null
+    }
+    if (areas == null) {
+        return true
+    }
+    if (this.size != areas.size) {
+        return true
+    }
+    return this.any { it.hasChanges(areas.find { area -> area.hankealueId == it.hankealueId }) }
```

Comment:
Checking against `hankealueId` seems like a good choice for finding changes.

Reply:
If the list ordering matters for tyoalueet, it should matter here as well.

Reply:
I got rid off these by overloading equals and hashCode in Tyoalue.